File: zilli/hybrid/executor.py
from __future__ import annotations

import logging
from dataclasses import dataclass, field
from typing import Optional

from zilli.hybrid.gatekeeper import ExecutionTarget, PrivacyGatekeeper
from zilli.models.base import GenerationResult
from zilli.models.registry import ModelRegistry
from zilli.privacy.engine import PrivacyVerdict

logger = logging.getLogger("zilli.hybrid.executor")


@dataclass
class HybridResult:
    text: str
    target: ExecutionTarget
    verdict: PrivacyVerdict
    model_name: str = ""
    tokens_in: int = 0
    tokens_out: int = 0
    error: Optional[str] = None
    warnings: list[str] = field(default_factory=list)


class HybridExecutor:
    def __init__(
        self,
        gatekeeper: PrivacyGatekeeper,
        registry: ModelRegistry,
    ):
        self.gatekeeper = gatekeeper
        self.registry = registry

    async def execute(
        self,
        prompt: str,
        tenant_id: str = "default",
        user_id: str = "",
        context_hint: str = "",
        preferred_cloud_provider: Optional[str] = None,
    ) -> HybridResult:
        from zilli.privacy.policy import CloudProvider

        provider = CloudProvider(preferred_cloud_provider) if preferred_cloud_provider else None
        decision = self.gatekeeper.decide(
            text=prompt,
            tenant_id=tenant_id,
            user_id=user_id,
            context_hint=context_hint,
            preferred_cloud=provider,
        )

        if decision.target == ExecutionTarget.REJECTED:
            logger.warning("Request rejected by gatekeeper: %s", decision.reason)
            return HybridResult(
                text="",
                target=ExecutionTarget.REJECTED,
                verdict=decision.verdict,
                error=f"Rejected by privacy gatekeeper: {decision.reason}",
                warnings=decision.warnings,
            )

        use_sanitized = decision.verdict.sanitized_text != prompt
        input_text = decision.verdict.sanitized_text if use_sanitized else prompt

        try:
            if decision.target == ExecutionTarget.CLOUD:
                result = await self._call_cloud(input_text, provider)
            elif decision.target == ExecutionTarget.LOCAL_WITH_CLOUD_FALLBACK:
                result = await self._call_local(input_text)
                if result.error:
                    logger.info("Local failed, falling back to cloud: %s", result.error)
                    result = await self._call_cloud(input_text, provider)
            else:
                result = await self._call_local(input_text)

            return HybridResult(
                text=result.text,
                target=decision.target,
                verdict=decision.verdict,
                model_name=result.model_name,
                tokens_in=result.tokens_in,
                tokens_out=result.tokens_out,
                error=result.error,
                warnings=decision.warnings,
            )
        except Exception as e:
            logger.error("Hybrid execution failed: %s", e)
            return HybridResult(
                text="",
                target=decision.target,
                verdict=decision.verdict,
                error=str(e),
                warnings=decision.warnings,
            )

    async def _call_local(self, prompt: str) -> GenerationResult:
        return await self.registry.generate_local(prompt)

    async def _call_cloud(self, prompt: str, provider: Optional[object] = None) -> GenerationResult:
        return await self.registry.generate_cloud(prompt, provider)

File: zilli-rs/src/hybrid/executor.rs
use serde::{Deserialize, Serialize};
use tracing::instrument;

use super::gatekeeper::{ExecutionTarget, PrivacyGatekeeper};
use crate::models::config::ModelRole;
use crate::models::registry::ModelRegistry;
use crate::privacy::classifier::DataClassifier;
use crate::privacy::consent::DataUse;
use crate::security::Sanitizer;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HybridResult {
    pub text: String,
    pub target: ExecutionTarget,
    pub verdict: String,
    pub model_name: String,
    pub tokens_in: u64,
    pub tokens_out: u64,
    pub error: Option<String>,
    pub warnings: Vec<String>,
}

const _CLOUD_MODEL_PREFIX: &str = "cloud-";

pub struct HybridExecutor {
    gatekeeper: PrivacyGatekeeper,
    model_registry: ModelRegistry,
    sanitizer: Sanitizer,
    classifier: DataClassifier,
}

impl HybridExecutor {
    pub fn new(
        gatekeeper: PrivacyGatekeeper,
        model_registry: ModelRegistry,
    ) -> Self {
        Self {
            gatekeeper,
            model_registry,
            sanitizer: Sanitizer::new(),
            classifier: DataClassifier::new(),
        }
    }

    #[instrument(skip(self, prompt))]
    pub async fn execute(
        &self,
        prompt: &str,
        model_role: ModelRole,
        data_use: &DataUse,
        tenant_id: &str,
    ) -> HybridResult {
        let classification = self.classifier.classify(prompt);

        let decision = self.gatekeeper.decide(
            prompt,
            classification.data_class,
            data_use,
            tenant_id,
        );

        match decision.target {
            ExecutionTarget::Local | ExecutionTarget::LocalWithCloudFallback => {
                let result = self
                    .model_registry
                    .generate(model_role.clone(), prompt, None, None)
                    .await;

                match result {
                    Ok(res) => HybridResult {
                        text: res.text,
                        target: decision.target,
                        verdict: decision.verdict.clone(),
                        model_name: res.model_name,
                        tokens_in: res.tokens_in,
                        tokens_out: res.tokens_out,
                        error: None,
                        warnings: decision.warnings.clone(),
                    },
                Err(e) => {
                        if decision.target == ExecutionTarget::LocalWithCloudFallback {
                            let sanitized = self.sanitizer.sanitize(prompt);
                            let fallback = self
                                .model_registry
                                .generate(model_role, &sanitized, None, None)
                                .await;

                            match fallback {
                                Ok(res) => HybridResult {
                                    text: res.text,
                                    target: ExecutionTarget::Cloud,
                                    verdict: "cloud_fallback".into(),
                                    model_name: res.model_name,
                                    tokens_in: res.tokens_in,
                                    tokens_out: res.tokens_out,
                                    error: None,
                                    warnings: decision.warnings,
                                },
                                Err(fb_e) => HybridResult {
                                    text: String::new(),
                                    target: ExecutionTarget::Rejected,
                                    verdict: "fallback_failed".into(),
                                    model_name: String::new(),
                                    tokens_in: 0,
                                    tokens_out: 0,
                                    error: Some(format!("Local failed: {}, Fallback failed: {}", e, fb_e)),
                                    warnings: decision.warnings,
                                },
                            }
                        } else {
                            HybridResult {
                                text: String::new(),
                                target: decision.target,
                                verdict: decision.verdict,
                                model_name: String::new(),
                                tokens_in: 0,
                                tokens_out: 0,
                                error: Some(e.to_string()),
                                warnings: decision.warnings,
                            }
                        }
                    }
                }
            }
            ExecutionTarget::Cloud => {
                let result = self
                    .model_registry
                    .generate(model_role, prompt, None, None)
                    .await;

                match result {
                    Ok(res) => HybridResult {
                        text: res.text,
                        target: ExecutionTarget::Cloud,
                        verdict: decision.verdict.clone(),
                        model_name: res.model_name,
                        tokens_in: res.tokens_in,
                        tokens_out: res.tokens_out,
                        error: None,
                        warnings: decision.warnings,
                    },
                    Err(e) => HybridResult {
                        text: String::new(),
                        target: ExecutionTarget::Rejected,
                        verdict: "execution_failed".into(),
                        model_name: String::new(),
                        tokens_in: 0,
                        tokens_out: 0,
                        error: Some(e.to_string()),
                        warnings: decision.warnings,
                    },
                }
            }
            ExecutionTarget::Rejected => HybridResult {
                text: String::new(),
                target: ExecutionTarget::Rejected,
                verdict: decision.verdict,
                model_name: String::new(),
                tokens_in: 0,
                tokens_out: 0,
                error: Some(decision.reason.clone()),
                warnings: decision.warnings,
            },
        }
    }
}
